flow/connectors/postgres/sink.go

```diff
@@ -0,0 +1,239 @@
+		if err != nil {
+			qe.logger.Error("[pg_query_executor] failed to process fetched rows", slog.Any("error", err))
+			return 0, err
+		}
+
+		qe.logger.Info(fmt.Sprintf("[pg_query_executor] fetched %d rows for query '%s'", numRows, query))
+		totalRecordsFetched += numRows
+
+		if numRows == 0 {
+			break
+		}
+	}
+
+	qe.logger.Info("Committing transaction")
+	if err := tx.Commit(ctx); err != nil {
+		qe.logger.Error("[pg_query_executor] failed to commit transaction", slog.Any("error", err))
+		err = fmt.Errorf("[pg_query_executor] failed to commit transaction: %w", err)
+		stream.Close(err)
+		return 0, err
+	}
+
+	qe.logger.Info(fmt.Sprintf("[pg_query_executor] committed transaction for query '%s', rows = %d",
+		query, totalRecordsFetched))
+	return totalRecordsFetched, nil
+}
+
+func (stream RecordStreamSink) CopyInto(ctx context.Context, _ *PostgresConnector, tx pgx.Tx, table pgx.Identifier) (int64, error) {
+	return tx.CopyFrom(ctx, table, stream.GetColumnNames(), model.NewQRecordCopyFromSource(stream.QRecordStream))
+}
+
+func (stream RecordStreamSink) GetColumnNames() []string {
+	return stream.Schema().GetColumnNames()
+}
+
+func (p PgCopyWriter) SetSchema(schema []string) {
+	if !p.schema.schemaSet {
+		p.schema.schema = schema
+		close(p.schema.schemaLatch)
+		p.schema.schemaSet = true
+	}
+}
+
+func (p PgCopyWriter) ExecuteQueryWithTx(
+	ctx context.Context,
+	qe *QRepQueryExecutor,
+	tx pgx.Tx,
+	query string,
+	args ...interface{},
+) (int, error) {
+	defer shared.RollbackTx(tx, qe.logger)
+
+	if qe.snapshot != "" {
+		_, err := tx.Exec(ctx, "SET TRANSACTION SNAPSHOT "+QuoteLiteral(qe.snapshot))
+		if err != nil {
+			qe.logger.Error("[pg_query_executor] failed to set snapshot",
+				slog.Any("error", err), slog.String("query", query))
+			err := fmt.Errorf("[pg_query_executor] failed to set snapshot: %w", err)
+			p.Close(err)
+			return 0, err
+		}
+	}
+
+	norows, err := tx.Query(ctx, query+" limit 0", args...)
+	if err != nil {
+		return 0, err
+	}
+
+	fieldDescriptions := norows.FieldDescriptions()
+	cols := make([]string, 0, len(fieldDescriptions))
+	for _, fd := range fieldDescriptions {
+		cols = append(cols, QuoteIdentifier(fd.Name))
+	}
+	p.SetSchema(cols)
+	norows.Close()
+
+	// TODO use pgx simple query arg parsing code (it's internal, need to copy)
+	for i, arg := range args {
+		var str string
+		switch arg := arg.(type) {
+		case nil:
+			str = "null"
+		case int64:
+			str = strconv.FormatInt(arg, 10)
+		case float64:
+			str = strconv.FormatFloat(arg, 'f', -1, 64)
+		case bool:
+			str = strconv.FormatBool(arg)
+		case string:
+			str = QuoteLiteral(arg)
+		case time.Time:
+			str = arg.Truncate(time.Microsecond).Format("'2006-01-02 15:04:05.999999999Z07:00:00'")
+		default:
+			return 0, fmt.Errorf("invalid arg type: %T", arg)
+		}
+		query = strings.ReplaceAll(query, fmt.Sprintf("$%d", i+1), str)
+	}
+
+	copyQuery := fmt.Sprintf("COPY (%s) TO STDOUT", query)
```

Comment:
no fetch size here as with the QValue type system, not sure if can lead to issues if sync isn't fast enough

Reply:
if sync is slow writes will block, as they should

will want to make a general benchmark at some point, even if it's only to have some numbers for a blog post